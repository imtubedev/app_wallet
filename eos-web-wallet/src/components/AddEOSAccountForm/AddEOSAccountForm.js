import * as React from "react";
import { Link } from "react-router-dom";
import { Field, reduxForm } from "redux-form";
import renderField from "components/Field";
import Button from "components/Button";

const AddEOSAccountForm = ({ callAPI, handleSubmit, submitting, match }) => (
  <article>
    <form onSubmit={handleSubmit(callAPI)} autoComplete="off">
      <h3>Connect your EOS Account</h3>
      <div className="box double">
        <div className="columns is-desktop">
          <div className="column is-half-desktop">
            <Field
              aria-describedby="account_name"
              className="input"
              id="account_name"
              label="EOS Account Name"
              name="account_name"
              required
              component={renderField}
              type="text"
              autoComplete="off"
            />
          </div>
        </div>
        <div className="columns is-desktop">
          <div className="column">
            <Field
              aria-describedby="active_key"
              className="input"
              id="active_key"
              label="EOS Active Key"
              name="active_key"
              required
              component={renderField}
              type="text"
              autoComplete="off"
            />
          </div>
          <div className="column">
            <Field
              aria-describedby="owner_key"
              className="input"
              id="owner_key"
              label="EOS Owner Key"
              name="owner_key"
              required
              component={renderField}
              type="text"
              autoComplete="off"
            />
          </div>
        </div>
        <div className="field is-grouped u-mt4">
          <div className="control">
            <Button
              disabled={submitting}
              className="button is-large is-primary"
              type="submit"
              text={submitting ? "Submitting..." : "Add Account"}
            />
          </div>
          <div className="control">
            <p>
              Donâ€™t have an account?{" "}
              <Link to="/accounts/create-eos">Sign Up</Link>
            </p>
          </div>
        </div>
      </div>
    </form>
  </article>
);

export default reduxForm({
  form: "connect-eos-account"
})(AddEOSAccountForm);
