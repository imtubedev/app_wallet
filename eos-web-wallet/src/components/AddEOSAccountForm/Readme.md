Connected EOS accounts example:

```js
const { Provider } = require( "react-redux");
const { Router } = require("react-router-dom");
const { createBrowserHistory } = require("history");
const store = {
  getState: () => ({}),
  dispatch: _ => _,
  subscribe: _ => _
};
const noop = () => {};

<Provider store={store}>
  <Router history={createBrowserHistory()}>
    <AddEOSAccountForm callAPI={noop} />
  </Router>
</Provider>
```
