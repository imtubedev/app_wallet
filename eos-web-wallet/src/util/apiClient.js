// import moment from "moment";
import { merge, uniq, map, flatMap } from "lodash";
import { Buffer } from "buffer";
import eos from "eosjs";

const { ecc, Fcbuffer } = eos.modules;

// Required at front of all transactions.
// NOTE: This is a default of 64 '0's, even though it's not implemented in eosd yet
const chainId = new Buffer("00".repeat(32), "hex");

const eosEndpoint =
  process.env.NODE_ENV === "test"
    ? process.env.REACT_APP_EOSD_TESTS_URI
    : process.env.REACT_APP_EOSD_URI;

const apiEndpoint =
  process.env.NODE_ENV === "test"
    ? process.env.REACT_APP_API_TESTS_URI
    : process.env.REACT_APP_API_URI;

const authHeader = {
  Authorization: `Bearer ${process.env.REACT_APP_API_KEY}`
};

const defaultOptions = {
  method: "GET",
  headers: {
    "Content-type": "application/json"
  }
};

const defaultAPIOptions = {
  method: "GET",
  headers: {
    "Content-type": "application/json",
    ...authHeader
  }
};
// Grab the correct, sorted scope from the list of messages
// intended for a transaction
const getScope = messages =>
  uniq(
    flatMap(messages, msg => [
      msg.code,
      ...map(msg.authorization, auth => auth.account)
    ])
  ).sort();

const rejectBadResponse = (response: Response): mixed =>
  response.ok || Promise.reject(response);

const request = async (
  path: string,
  options: mixed = {},
  mergeDefaults: boolean = true
) => {
  const opts = mergeDefaults ? merge({}, defaultOptions, options) : options;
  try {
    const resp = await fetch(path, opts);
    console.log("<resp>", resp);
    await rejectBadResponse(resp);
    // only attempt to parse JSON if proper content-type
    if (resp.headers && resp.headers.get("content-type").indexOf("json") > -1)
      return resp.json();
    return resp.text();
  } catch (error) {
    console.log(error);
    if (typeof error === "string") {
      return Promise.reject(error);
    }
    if (error instanceof Error) {
      return Promise.reject(error);
    }
    const data =
      error.headers && error.headers.get("content-type").indexOf("json") > -1
        ? await error.json()
        : await error.text();
    return Promise.reject(data);
  }
};

const configureClient = (target, opts = {}) => {
  if (!opts.keyProvider)
    throw new Error(`api client connection needs a keyProvider`);

  const eosNode = eos.Testnet({
    keyProvider: opts.keyProvider,
    httpEndpoint: eosEndpoint,
    debug: process.env.NODE_ENV !== "production",
    broadcast: false
  });

  return Object.assign(target, {
    keyProvider: opts.keyProvider,
    eos: eosNode
  });
};

class APIClient {
  constructor(opts) {
    configureClient(this, opts);
    this.structs = this.eos.fc.structs;
  }

  setKeyProvider(keyProvider) {
    this.keyProvider = keyProvider;
    console.log("--> set keyProvider:", this.keyProvider);
  }

  async get(path, opts = {}) {
    const options = Object.assign({}, defaultAPIOptions, opts);
    return await request(`${apiEndpoint}${path}`, options);
  }

  async post(path, data, opts = {}) {
    const options = Object.assign({}, defaultAPIOptions, opts, {
      method: "POST",
      headers: authHeader,
      body: JSON.stringify(data)
    });
    return await request(`${apiEndpoint}${path}`, options);
  }

  async postTransaction(path, data, opts = {}) {
    let { scope, messages } = data;

    if (!messages) {
      throw new Error("APIClient.post() data missing `messages` property");
    }
    if (!Array.isArray(messages)) {
      throw new Error(
        "APIClient.post() messages should be an array of one or more messages"
      );
    }
    if (!scope) {
      // Attempt to get scope from messages
      scope = getScope(messages);
    }
    // ensure scope is unique and sorted ascending alphabetically
    scope = uniq(scope).sort();

    // build a public->private key mapping
    const keyMap = this.keyProvider.reduce((map, wif) => {
      map[ecc.privateToPublic(wif)] = wif;
      return map;
    }, {});

    try {
      // Get info on head on chain
      console.log(eosEndpoint);
      const info = await request(`${eosEndpoint}/v1/chain/get_info`);
      const head = new Date(`${info.head_block_time}Z`);
      head.setSeconds(head.getSeconds() + 60);
      const expr = head.toISOString().split(".")[0];
      // const expr = moment(new Date(`${info.head_block_time}Z`))
      //   .add(60, "seconds")
      //   .toISOString()
      //   .split(".")[0];

      // Build transaction from data
      const transaction = {
        ref_block_num: info.head_block_num & 0xffff,
        ref_block_prefix: new Buffer(info.head_block_id, "hex").readUInt32LE(8),
        expiration: expr,
        scope,
        read_scope: [],
        messages
      };
      // construct a buffer of the transaction to sign
      const trBuffer = Buffer.concat([
        chainId,
        Fcbuffer.toBuffer(this.structs.transaction, transaction)
      ]);

      const { required_keys } = await this.eos.getRequiredKeys(
        transaction,
        Object.keys(keyMap)
      );

      const signatures = [];
      (required_keys || []).forEach(pubKey => {
        if (!keyMap[pubKey]) {
          throw new Error(
            `APIClient: missing necessary private key for signing, needs private key for ${
              pubKey
            }`
          );
        }
        signatures.push(ecc.sign(trBuffer, keyMap[pubKey]));
      });

      transaction.signatures = signatures;

      return await this.post(path, transaction, opts);
    } catch (err) {
      console.error("fetch error:", err);
      return Promise.reject(err.statusText || err);
    }
  }
}

// Export singleton instance
export let apiClient = new APIClient({ keyProvider: [] });

// Export class
export default APIClient;
