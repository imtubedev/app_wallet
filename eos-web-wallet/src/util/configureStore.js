import { createStore, compose, combineReducers, applyMiddleware } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import { reducer as form } from "redux-form";
import { routerMiddleware } from "react-router-redux";
import { routerReducer } from "react-router-redux";
import app from "redux-modules/app";
import user from "redux-modules/user";
import usersList from "redux-modules/users-list";
import eosBalances from "redux-modules/eos-balances";
import notification from "redux-modules/notifications";
import transactions from "redux-modules/transactions";
import profile from "redux-modules/profile";
import { createMemoryHistory } from "history";
import middlewares from "middleware";
import sessionStorage from "redux-persist/lib/storage/session";
import { rehydrateAccounts } from "../middleware/account-persist/account-persist-actions";
import { selectWalletUserAuthenticated } from "../redux-modules/user/user-selectors";

let store;
let persistor;

const reducers = {
  app,
  form: persistReducer(
    {
      key: "forms",
      storage: sessionStorage
    },
    form
  ),
  user,
  usersList,
  eosBalances,
  notification,
  profile,
  transactions,
  routing: routerReducer
};

export const configureStore = (
  preloadState,
  history = createMemoryHistory()
) => {
  /* eslint-disable-next-line no-underscore-dangle */
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(...middlewares, routerMiddleware(history))
  );

  store = createStore(combineReducers(reducers), preloadState, enhancers);
  store.reducers = reducers;
  persistor = persistStore(store, undefined, () => {
    if (selectWalletUserAuthenticated(store.getState())) {
      store.dispatch(rehydrateAccounts());
    }
  });

  return { store, persistor };
};

export const addReducers = reducers => {
  let added = false;
  reducers.forEach(({ name, reducer }) => {
    if (!store.reducers[name]) {
      store.reducers = {
        ...store.reducers,
        [name]: reducer
      };
      added = true;
    }
  });

  if (added) {
    console.log(JSON.stringify(store.reducers));
    store.replaceReducer(combineReducers(store.reducers));
    persistor.persist();
  }
};

export const removeReducers = reducerNames => {
  let removed = false;
  reducerNames.forEach(reducerName => {
    if (store.reducers[reducerName]) {
      Reflect.deleteProperty(store.reducers, reducerName);
      removed = true;
    }
  });

  if (removed) {
    console.log(JSON.stringify(store.reducers));
    store.replaceReducer(combineReducers(store.reducers));
  }
};
