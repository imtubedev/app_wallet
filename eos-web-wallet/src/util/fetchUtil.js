//@flow
/* global fetch */

export const rejectBadResponse = (response: Response): mixed =>
  response.ok || Promise.reject(response);

const defaultAppOptions = {
  method: "GET",
  credentials: "include",
  headers: {
    "Content-Type": "application/json"
  }
};

export const appRequest = async (path: string, options: mixed = {}) => {
  try {
    const resp = await fetch(path, {
      ...defaultAppOptions,
      ...options
    });
    await rejectBadResponse(resp);
    return resp.json();
  } catch (error) {
    return Promise.reject(await error.json());
  }
};

const apiHost: string = process.env.REACT_APP_API_URI || "";
const apiKey: string = process.env.REACT_APP_API_KEY || "";
const defaultApiOptions = {
  method: "GET",
  headers: {
    "Content-type": "application/json",
    Authorization: `Bearer ${apiKey}`
  }
};

export const apiRequest = async (path: string, options: mixed = {}): any => {
  let resp;
  try {
    resp = await fetch(`${apiHost}${path}`, {
      ...defaultApiOptions,
      ...options
    });
    await rejectBadResponse(resp);
    // just grab as text (rather than check response headers for content type)
    let text = await resp.text();
    try {
      // attempt to parse as JSON
      return JSON.parse(text);
    } catch (error) {
      // Not json, format and return
      return {
        status: resp.status,
        text: resp.statusText,
        message: text
      };
    }
  } catch (error) {
    return Promise.reject(await error.json());
  }
};
