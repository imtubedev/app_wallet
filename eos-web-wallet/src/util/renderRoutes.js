import React from 'react';
import Switch from 'react-router/Switch';
import Route from 'react-router/Route';
import { withRouter } from 'react-router-dom';

const renderRoutes = (routes, extraProps = {}, switchProps = {}) => (
  routes ? (
    <Switch {...switchProps}>
      {routes.map((route, i) => {
        const RouteComponent = withRouter(route.component);
        return (
          <Route
            key={route.key || i}
            path={route.path}
            exact={route.exact}
            strict={route.strict}
            render={props => <RouteComponent {...props} {...extraProps} route={route} routes={route.routes} />}
          />);
      })}
    </Switch>
  ) : null
);

export default renderRoutes;
