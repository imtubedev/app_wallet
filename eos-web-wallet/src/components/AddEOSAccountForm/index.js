import AddEOSAccountForm from "./AddEOSAccountForm";
import { connect } from "react-redux";
import { addEOSAccount } from "../../thunks/eos-account";
import { withRouter } from "react-router-dom";

/* eslint-disable camelcase */
const mapDispatchToProps = dispatch => ({
  callAPI(values) {
    return dispatch(
      addEOSAccount(values.account_name, values.owner_key, values.active_key)
    );
  }
});

const ConnectedAddEOSAccountForm = withRouter(
  connect(null, mapDispatchToProps)(AddEOSAccountForm)
);

export default ConnectedAddEOSAccountForm;
