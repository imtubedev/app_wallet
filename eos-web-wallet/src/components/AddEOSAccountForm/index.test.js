import * as React from "react";
import { Provider } from "react-redux";
import { StaticRouter } from "react-router";
const { createBrowserHistory } = require("history");
import { configureStore } from "util/configureStore";
import AddEOSAccountForm from "./";

describe("<AddEOSAccountForm /> container", () => {
  it("renders without crashing", () => {
    const { store } = configureStore({});
    snapshot(
      <Provider store={store}>
        <StaticRouter location="/" context={{}}>
          <AddEOSAccountForm />
        </StaticRouter>
      </Provider>
    );
  });
});
