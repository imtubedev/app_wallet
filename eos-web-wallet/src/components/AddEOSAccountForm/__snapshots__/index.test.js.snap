// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddEOSAccountForm /> container renders without crashing 1`] = `
<article>
  <form
    autoComplete="off"
    onSubmit={[Function]}
  >
    <h3>
      Connect your EOS Account
    </h3>
    <div
      className="box double"
    >
      <div
        className="columns is-desktop"
      >
        <div
          className="column is-half-desktop"
        >
          <div
            className="field"
          >
            <label
              className="label"
              htmlFor="account_name"
            >
              EOS Account Name
            </label>
            <div
              className="control "
            >
              <input
                aria-describedby="account_name"
                autoComplete="off"
                className="input"
                id="account_name"
                name="account_name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="columns is-desktop"
      >
        <div
          className="column"
        >
          <div
            className="field"
          >
            <label
              className="label"
              htmlFor="active_key"
            >
              EOS Active Key
            </label>
            <div
              className="control "
            >
              <input
                aria-describedby="active_key"
                autoComplete="off"
                className="input"
                id="active_key"
                name="active_key"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="column"
        >
          <div
            className="field"
          >
            <label
              className="label"
              htmlFor="owner_key"
            >
              EOS Owner Key
            </label>
            <div
              className="control "
            >
              <input
                aria-describedby="owner_key"
                autoComplete="off"
                className="input"
                id="owner_key"
                name="owner_key"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="field is-grouped u-mt4"
      >
        <div
          className="control"
        >
          <button
            className="button button is-large is-primary"
            onClick={undefined}
          >
            <span>
              Add Account
            </span>
          </button>
        </div>
        <div
          className="control"
        >
          <p>
            Donâ€™t have an account?
             
            <a
              href="/accounts/create-eos"
              onClick={[Function]}
            >
              Sign Up
            </a>
          </p>
        </div>
      </div>
    </div>
  </form>
</article>
`;
